File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2023/Y23D12.kt
package org.gristle.pdxpuzzles.advent.y2023

import org.gristle.pdxpuzzles.advent.utilities.Day

class Y23D12(input: String) : Day {
    data class SpringRow(val conditions: String, val damageReport: List<Int>) {
        // expands the SpringRow in accordance with Part 2 rules
        fun expand(): SpringRow = SpringRow(
            List(5) { conditions }.joinToString("?"),
            List(5) { damageReport }.flatten()
        )
        
        // cache used for DP function 'arrangements,' below. 
        private val cache = mutableMapOf<Pair<Int, Int>, Long>()
        
        // DP function that uses state objects to delve deeper and deeper into the string using DFS, until a base case 
        // is found. Cache is updated along the way, so that subsequent dives down the string become instant as soon as
        // there is any repetition of remaining state.
        fun arrangements(
            conditionsIndex: Int = 0,
            damageReportIndex: Int = 0
        ): Long = cache.getOrPut(conditionsIndex to damageReportIndex) {
            // Base case. Takes states that have assigned broken springs corresponding to the entire damage report 
            // and returns 1 if valid, 0 if invalid. Valid states are those with no remaining '#' in the string.
            if (damageReportIndex == damageReport.size) {
                val damagedSpringRemaining = (conditionsIndex..conditions.lastIndex).any { conditions[it] == '#' }
                return@getOrPut if (damagedSpringRemaining) 0 else 1
            }
            // Otherwise, we go recursive by trying to fit the next group of broken springs in every place along the 
            // block. This starts as a sequence of indexes, from 0 until the length of the block minus the 
            // fulfillment size (to account for the size of the fulfillment itself in the string).
            val brokenGroup = damageReport[damageReportIndex]
            (conditionsIndex..conditions.length - brokenGroup)
                .asSequence()
                // stop the sequence if a '#' precedes the index b/c '#' cannot be skipped
                .takeWhile { index -> conditions.getOrNull(index - 1) != '#' }
                .filter { index ->
                    // filter out invalid placements, in cascading fashion
                    // if the placement includes a '.', invalid b/c '.' means not broken
                    // if the placement has no part of the string after it, valid b/c nothing else to consider
                    // if the character following the placement is '#', invalid b/c that extra '#' would overfulfill
                    // otherwise valid
                    when {
                        (index until index + brokenGroup).any { conditions[it] == '.' } -> false
                        index + brokenGroup == conditions.length -> true
                        conditions[index + brokenGroup] == '#' -> false
                        else -> true
                    }
                }.sumOf { index -> arrangements(index + brokenGroup + 1, damageReportIndex + 1) }
        }
    }

    // parsing input to SpringRows
    private val springRows: List<SpringRow> = input.lines().map { line ->
        val (conditions, damageReportStr) = line.split(' ')
        val damageReport: List<Int> = damageReportStr.split(',').map(String::toInt)
        SpringRow(conditions, damageReport)
    }

    override fun part1(): Long = springRows.sumOf(SpringRow::arrangements)
    
    override fun part2(): Long = springRows.map(SpringRow::expand).sumOf(SpringRow::arrangements)
}

fun main() = Day.runDay(Y23D12::class)

//    Class creation: 17ms
//    Part 1: 7344 (18ms)
//    Part 2: 1088006519007 (135ms)
//    Total time: 171ms

@Suppress("unused")
private val sampleInput = listOf(
    """???.### 1,1,3
.??..??...?##. 1,1,3
?#?#?#?#?#?#?#? 1,3,1,6
????.#...#... 4,1,1
????.######..#####. 1,6,5
?###???????? 3,2,1
"""
)

File: rust/advent/src/bin/2023/y23d12.rs
use std::iter;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rustc_hash::FxHashMap;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Vec<(String, Vec<usize>)>;
type Output = usize;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(23, 12).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug)]
struct SpringRow {
    conditions: String,
    damage_report: Vec<usize>,
    cache: FxHashMap<State, usize>,
}

impl SpringRow {
    fn arrangements(&mut self, s: State) -> usize {
        if let Some(value) = self.cache.get(&s) {
            *value
        } else {
            // do not consider conditions already handled in previous states
            // if state place exceeds the conditions length string, we are done and the block is blank
            let block = if s.conditions_index < self.conditions.len() {
                self.conditions[s.conditions_index..].to_string()
            } else {
                String::from("")
            };

            // the # of consecutive broken springs in the damage report that we try to place along the row
            let fulfillment = self.damage_report.get(s.damage_index).unwrap_or(&0).to_owned();

            // Base case. Takes states that have fulfilled the entire damage report and returns 1 if valid,
            // 0 if invalid. Valid states are those with no remaining '#' in the current or any future blocks,
            // and that have filled all the damaged spring requirements
            if fulfillment == 0usize {
                let value = if block.find('#').is_some() {
                    0
                } else {
                    1
                };
                self.cache.insert(s, value);
                return value;
            }

            // Otherwise, we go recursive by trying to fit the fulfillment in every place along the block
            // This starts as a sequence of indexes, from 0 until the length of the block minus the fulfillment size
            // (to account for the size of the fulfillment itself in the string).
            let value = if block.len() >= fulfillment {
                (0..=block.len() - fulfillment)
                    .take_while(|index| { *index == 0 || block.as_bytes()[index - 1] as char != '#' })
                    .filter(|index| {
                        // filter out invalid placements, in cascading fashion
                        // if the placement includes a '.', invalid b/c '.' means not broken
                        // if the placement has no part of the string after it, valid b/c nothing else to consider
                        // if the character following the placement is '#', invalid b/c that extra '#' would overfulfill
                        // otherwise valid
                        if let Some(_) = &block[*index..index + fulfillment].find('.') {
                            false
                        } else if index + fulfillment == block.len() {
                            true
                        } else if block.as_bytes()[index + fulfillment] == '#' as u8 {
                            false
                        } else {
                            true
                        }
                    })
                    .map(|index| {
                        let new_state = State {
                            conditions_index: s.conditions_index + index + fulfillment + 1,
                            damage_index: s.damage_index + 1,
                        };
                        self.arrangements(new_state)
                    })
                    .sum()
            } else {
                0
            };
            self.cache.insert(s, value);
            value
        }
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
struct State {
    conditions_index: usize,
    damage_index: usize,
}

fn parse_input(input: &str) -> Input {
    input.lines()
    .map(|line| {
        let (conditions, damage_str) = line.split_once(' ').unwrap();
        let damage_report: Vec<usize> = damage_str.split(',')
            .map(|s| s.parse::<usize>().unwrap())
            .collect();
        (conditions.to_string(), damage_report)
    })
    .collect()
}

fn solve(spring_rows: Vec<SpringRow>) -> usize {
    spring_rows.into_iter()
        .map(|mut spring_row| {
            let result = spring_row.arrangements(
                State {
                    conditions_index: 0,
                    damage_index: 0,
                }
            );
            result
        }).sum()
}

fn part1(spring_reports: &Input) -> Output {
    let spring_rows: Vec<SpringRow> = spring_reports.iter()
        .map(|(conditions, damage_report)| {
            let cache: FxHashMap<State, usize> = FxHashMap::default();
            SpringRow { 
                conditions: conditions.clone(), 
                damage_report: damage_report.clone(), 
                cache 
            }
        })
        .collect();

    solve(spring_rows)
}


fn part2(spring_reports: &Input) -> Output {
    let spring_rows: Vec<SpringRow> = spring_reports.into_iter()
        .map(|(conditions, damage_report)| {
            let expanded_conditions = iter::repeat(conditions).take(5)
                .join("?");
            let expanded_damage_report = damage_report.repeat(5);
            let cache: FxHashMap<State, usize> = FxHashMap::default();
            SpringRow {
                conditions: expanded_conditions,
                damage_report: expanded_damage_report,
                cache,
            }
        })
        .collect();

    solve(spring_rows)
}

#[test]
fn default() {
    let input = get_input(23, 12).unwrap();
    let input = parse_input(&input);
    assert_eq!(7344, part1(&input));
    assert_eq!(1088006519007, part2(&input));
}

// Input parsed (174μs)
// 1. 7344 (873μs)
// 2. 1088006519007 (17ms)
// Total: 18ms